frontend/src/scenes/settings/environment/CustomChannelTypes.tsx

```diff
@@ -0,0 +1,319 @@
+            .map((channelType) => ({ label: channelType, key: channelType }))
+    }, [customChannelTypeRules])
+
+    return (
+        <div>
+            <p>
+                You can create custom channel types by defining rules that match incoming events. The first matching
+                rule is used, and if no rule matches (or if none are defined) then the{' '}
+                <Link to="https://posthog.com/docs/data/channel-type#channel-type-calculation">
+                    default channel type
+                </Link>{' '}
+                is used.
+            </p>
+            <p>
+                To debug, try the{' '}
+                <Link to={urls.sessionAttributionExplorer()} target="_blank">
+                    session attribution explorer tool
+                </Link>
+            </p>
+            <ChannelTypeEditor
+                handleChange={setCustomChannelTypeRules}
+                initialCustomChannelTypeRules={customChannelTypeRules}
+                channelTypeOptions={channelTypeOptions}
+                onSave={() => {
+                    updateCurrentTeam({
+                        modifiers: { customChannelTypeRules: sanitizeCustomChannelTypeRules(customChannelTypeRules) },
+                    })
+                    reportCustomChannelTypeRulesUpdated(customChannelTypeRules.length)
+                    setSavedCustomChannelTypeRules(customChannelTypeRules)
+                }}
+                isSaveDisabled={isEqual(customChannelTypeRules, savedCustomChannelTypeRules)}
+            />
+        </div>
+    )
+}
+
+export interface ChannelTypeEditorProps {
+    handleChange: (rules: CustomChannelRule[]) => void
+    initialCustomChannelTypeRules: CustomChannelRule[]
+    channelTypeOptions: LemonInputSelectOption[]
+    isSaveDisabled: boolean
+    onSave: () => void
+}
+
+export function ChannelTypeEditor({
+    handleChange,
+    initialCustomChannelTypeRules,
+    channelTypeOptions,
+    isSaveDisabled,
+    onSave,
+}: ChannelTypeEditorProps): JSX.Element {
+    return (
+        <VerticalNestedDND<CustomChannelCondition, CustomChannelRule>
+            initialItems={initialCustomChannelTypeRules}
+            renderContainerItem={(rule, { updateContainerItem }) => {
+                return (
+                    <div className="flex flex-col space-y-2">
+                        <div className="flex flex-row items-center space-x-2">
+                            <span>Set Channel type to</span>
+                            <LemonInputSelect
+                                className="flex-1"
+                                mode="single"
+                                allowCustomValues={true}
+                                value={[rule.channel_type]}
+                                onChange={(channelType) =>
+                                    updateContainerItem({
+                                        ...rule,
+                                        channel_type: channelType[0],
```

Comment:
Quick question about the `LemonInputSelect`. If we're saying its mode is `single`, why is it returning an array? Should we change that?

Reply:
In `single` mode it returns an array with 1 item. This makes it easier to switch between single and multiple, and makes the types easier. (I'm kinda guessing about this, I didn't write the LemonInputSelect component, but I think it's makes sense as is).